mirrord-layer/src/sockets.rs

```diff
@@ -1,240 +1,373 @@
+                error!("socket is in invalid state for bind {:?}", socket.state);
+                return libc::EINVAL;
+            }
+            Some(socket) => socket,
+            None => {
+                debug!("bind: no socket found for fd: {}", &sockfd);
+                return unsafe { libc::bind(sockfd, addr, addrlen) };
+            }
+        }
+    };
+
+    let raw_addr = unsafe { OsSocketAddr::from_raw_parts(addr as *const u8, addrlen as usize) };
+    let parsed_addr = match raw_addr.into_addr() {
+        Some(addr) => addr,
+        None => {
+            error!("bind: failed to parse addr");
+            return libc::EINVAL;
         }
+    };
+
+    debug!("bind:port: {}", parsed_addr.port());
+    if is_ignored_port(parsed_addr.port()) {
+        debug!("bind: ignoring port: {}", parsed_addr.port());
+        return unsafe { libc::bind(sockfd, addr, addrlen) };
     }
 
-    pub fn open_connection(&self, connection_id: ConnectionId, port: Port) {
-        let mut connections = self.connections.lock().unwrap();
-        if let Some(mut socket) = connections.remove_alt(&port) {
-            debug!("new connection id: {:?}", connection_id);
-            let mut queues = self.connection_queues.lock().unwrap();
-            match queues.get_mut(&socket.read_fd) {
-                Some(queue) => {
-                    queue.add(connection_id).unwrap();
+    socket.state = SocketState::Bound(Bound {
+        address: parsed_addr,
+    });
+
+    let mut state = STATE.lock().unwrap();
+    state.sockets.insert(socket);
+    0
+}
+
+unsafe extern "C" fn bind_detour(
+    sockfd: c_int,
+    addr: *const sockaddr,
+    addrlen: socklen_t,
+) -> c_int {
+    bind(sockfd, addr, addrlen)
+}
+
+fn listen(sockfd: RawFd, _backlog: c_int) -> c_int {
+    debug!("listen called");
+    let mut socket = {
+        let mut state = STATE.lock().unwrap();
+        match state.sockets.take(&sockfd) {
+            Some(socket) => socket,
+            None => {
+                debug!("listen: no socket found for fd: {}", &sockfd);
+                return unsafe { libc::listen(sockfd, _backlog) };
+            }
+        }
+    };
+    match socket.state {
+        SocketState::Bound(bound) => {
+            socket.state = SocketState::Listening;
+            let mut os_addr = match socket.domain {
+                libc::AF_INET => {
+                    OsSocketAddr::from(SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 0))
                 }
+                libc::AF_INET6 => {
+                    OsSocketAddr::from(SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0))
+                }
+                _ => {
+                    // shouldn't happen
+                    debug!("unsupported domain");
+                    return libc::EINVAL;
+                }
+            };
+
+            let ret = unsafe { libc::bind(sockfd, os_addr.as_ptr(), os_addr.len()) };
```

Comment:
So does this call the hook, not find the socket, then call the actual function?

Reply:
yup